kubernetes/src/apiClient.c

```diff
@@ -452,6 +456,10 @@ size_t writeDataCallback(void *buffer, size_t size, size_t nmemb, void *userp) {
     apiClient->dataReceived = (char *)realloc( apiClient->dataReceived, apiClient->dataReceivedLen + size_this_time + 1);
     memcpy(apiClient->dataReceived + apiClient->dataReceivedLen, buffer, size_this_time);
     apiClient->dataReceivedLen += size_this_time;
+    ((char*)apiClient->dataReceived)[apiClient->dataReceivedLen] = '\0'; // the space size of (apiClient->dataReceived) = dataReceivedLen + 1
+    if (apiClient->watch_func) {
```

Comment:
I think the way to do this without modifying the API client is to split the api invoke into two phases:

1) Prepare the call
2) Execute the call

That way the watch code can run #1 from the generated code, but then take over the call to get access to the raw stream.

It's not going to be maintainable long term to add this to the generated code as it is Kubernetes specific.

Reply:
Hi @brendandburns 

Do you mean splitting the generator function ```apiClient_invoke``` to 2 parts ?

I do not find a way to "take over" the call to get access to the raw stream in kubernetes-client/c because ```curl_easy_perform``` in ```apiClient_invoke``` of ```watch``` request  will never return to caller. So I have to add the callback function into ```writeDataCallback```

I think ```apiClient->watch_func``` is not Kubernetes specific, other swagger client can use this function to support ```watch```

Reply:
Yes, I'm suggesting breaking `apiClient_invoke` into two parts:

```c
// the whole invoke function, currently a single function is split into two parts.
apiClient_invoke(...) {
   // Get the curl call ready
   apiClient_prepare(...);

   // actually make the curl call
   apiClient_call(...);
}

// Can be called separately to just prepare but not execute a call.
apiClient_prepare(...) {}

// Probably not used by users.
apiClient_call(...) {}
```

Reply:
Note that this separation will also be useful for port-forward/logs/etc which don't follow REST standards.

Reply:
Hi @brendandburns 

After splitting ```apiClient_invoke``` to ```apiClient_prepare``` and ```apiClient_call```,   how can kubernetes-client/c take over to access to the raw stream ?  

In current apiClient_invoke
```
        res = curl_easy_perform(handle);   // <--- curl_easy_perform will never return when "watch"
```

Reply:
Ok, I see the issue now.

If we change `watch_func` to `data_callback_func` or some-such to make it more generic, then I think it will be to ok to change the client as you suggest.

But we may want to switch the whole client over to `curl_multi_perform` instead eventually.

Reply:
how about using curl_multi_perform instead of using curl_easy_perform?
curl_easy_perform will be blocked and never return.

Reply:
Suppose to use ```curl_multi_perform``` in ```apiClient_call()```

```c
apiClient_call(...) {
...
    do {
        int numfds=0;
        int res = curl_multi_wait(cm, NULL, 0, MAX_WAIT_MSECS, &numfds);
        if(res != CURLM_OK) {
            fprintf(stderr, "error: curl_multi_wait() returned %d\n", res);
            return EXIT_FAILURE;
        }
        curl_multi_perform(cm, &still_running);
    } while(still_running);
...
}
```

How does the watch code take over to access the raw stream ?

Reply:
@clearday4

> I agree. how about using curl_multi_perform instead of using curl_easy_perform?
> or set timeout like curl_easy_setopt(curl_, CURLOPT_TIMEOUT, timeout_)
> currently curl_easy_perform will be blocked and never return.

Setting timeout is not a solution,  after curl_easy_perform returns after timeout, the http connection will disconnect,  watch events will never come, this is not right for the watch function.

Reply:
Hi @brendandburns 

I tried to implement with ```curl_multi_* ```, but I found it still need inject a function to ```apiClient_call``` to access the raw stream, it will do more changes to api client.

For the ```watch```, because we issue only one request at one time,  ```curl_multi_*``` is not better than ```curl_easy_*``` ,  the function ```curl_multi_wait``` will return as long as there is data comming,  so we need poll in a loop ```while(still_running)```

So I suggest adopting my current solution: 
1. Only add a callback function pointer```data_callback_func```, it does small change to api client. 
2. Other functions in watch/watch_util.{h,c} are committed to this repo (kubernetes-client/c)

If somebody has time to try a new solution with ```curl_multi_*````, we can update/change this part in the future.

cc @clearday4

Reply:
Your suggestions look good to me. Let's do it!